File: solar_mqtt/solar/discovery.py
"""Home Assistant MQTT Discovery（Feature #2）。

發佈 `homeassistant/sensor/<object_id>/config` 的 retained JSON，HA 自動建立感測器。
摘要：每廠 3 個（總功率、今日、本月）
各區：每區 6 個（功率、今日、本月、累計、容量、有效時數）

設計原則：
- 一次性在啟動（或 zone 清單首次獲得）後發佈，重複發也沒關係（HA 會覆蓋）
- value_template 從 {prefix}/{factory}/... 的 JSON payload 取 value 欄位
- unique_id 用 factory + zone_id + metric，跨廠不衝突
- device 分組：每廠一個 device
"""

from __future__ import annotations

import json
from typing import Callable

PublishFn = Callable[[str, str, bool], bool]  # (topic, payload, retain)


def _device(factory_id: str) -> dict:
    return {
        "identifiers": [f"ez_solar_{factory_id}"],
        "name": f"EZ-Solar {factory_id}",
        "manufacturer": "Joseph-Tech / 國瑞汽車",
        "model": "EZ-Solar Dashboard",
    }


def _sensor_config(
    *,
    factory_id: str,
    object_id: str,
    name: str,
    state_topic: str,
    unit: str | None,
    device_class: str | None,
    state_class: str | None = "measurement",
    value_template: str = "{{ value_json.value }}",
    icon: str | None = None,
) -> dict:
    cfg: dict = {
        "name": name,
        "unique_id": object_id,
        "object_id": object_id,
        "state_topic": state_topic,
        "value_template": value_template,
        "device": _device(factory_id),
    }
    if unit:
        cfg["unit_of_measurement"] = unit
    if device_class:
        cfg["device_class"] = device_class
    if state_class:
        cfg["state_class"] = state_class
    if icon:
        cfg["icon"] = icon
    return cfg


def publish_factory_discovery(
    publish: PublishFn,
    *,
    ha_prefix: str,
    prefix: str,
    factory_id: str,
    zones: list[dict],
) -> None:
    base = f"{prefix}/{factory_id}"

    entries: list[tuple[str, dict]] = []

    # Summary
    entries.append(
        (
            f"ez_solar_{factory_id}_total_power",
            _sensor_config(
                factory_id=factory_id,
                object_id=f"ez_solar_{factory_id}_total_power",
                name=f"{factory_id} 總功率",
                state_topic=f"{base}/total_power_kw",
                unit="kW",
                device_class="power",
            ),
        )
    )
    entries.append(
        (
            f"ez_solar_{factory_id}_today_energy",
            _sensor_config(
                factory_id=factory_id,
                object_id=f"ez_solar_{factory_id}_today_energy",
                name=f"{factory_id} 今日發電",
                state_topic=f"{base}/today_mwh",
                unit="MWh",
                device_class="energy",
                state_class="total_increasing",
            ),
        )
    )
    entries.append(
        (
            f"ez_solar_{factory_id}_month_energy",
            _sensor_config(
                factory_id=factory_id,
                object_id=f"ez_solar_{factory_id}_month_energy",
                name=f"{factory_id} 本月發電",
                state_topic=f"{base}/month_mwh",
                unit="MWh",
                device_class="energy",
                state_class="total_increasing",
            ),
        )
    )

    # Per-zone
    for z in zones:
        zid = z["zone_id"]
        zname = z.get("name") or f"Zone{zid}"
        ztopic = f"{base}/zone/{zid}"
        obj_prefix = f"ez_solar_{factory_id}_zone{zid}"

        metrics = [
            ("power", "功率", "power_kw", "kW", "power", "measurement"),
            ("today", "今日發電", "today_kwh", "kWh", "energy", "total_increasing"),
            ("month", "本月發電", "month_mwh", "MWh", "energy", "total_increasing"),
            ("total", "累計發電", "total_mwh", "MWh", "energy", "total_increasing"),
            ("capacity", "裝置容量", "capacity_kwp", "kWp", None, "measurement"),
            ("hours", "今日有效時數", "today_hours", "h", None, "measurement"),
        ]
        for suffix, zh, topic_key, unit, dev_cls, state_cls in metrics:
            entries.append(
                (
                    f"{obj_prefix}_{suffix}",
                    _sensor_config(
                        factory_id=factory_id,
                        object_id=f"{obj_prefix}_{suffix}",
                        name=f"{factory_id} {zname} {zh}",
                        state_topic=f"{ztopic}/{topic_key}",
                        unit=unit,
                        device_class=dev_cls,
                        state_class=state_cls,
                    ),
                )
            )

    for object_id, cfg in entries:
        topic = f"{ha_prefix}/sensor/{object_id}/config"
        publish(topic, json.dumps(cfg, ensure_ascii=False), True)

    print(f"[{factory_id}] HA Discovery 已發佈 {len(entries)} 個 entity")


def remove_factory_discovery(
    publish: PublishFn,
    *,
    ha_prefix: str,
    factory_id: str,
    zone_ids: list[int],
) -> None:
    """清除該廠所有 discovery（發空 payload 給 HA 會自動移除 entity）。"""
    ids = [
        f"ez_solar_{factory_id}_total_power",
        f"ez_solar_{factory_id}_today_energy",
        f"ez_solar_{factory_id}_month_energy",
    ]
    for zid in zone_ids:
        for suffix in ("power", "today", "month", "total", "capacity", "hours"):
            ids.append(f"ez_solar_{factory_id}_zone{zid}_{suffix}")
    for object_id in ids:
        publish(f"{ha_prefix}/sensor/{object_id}/config", "", True)

File: solar_mqtt_go/internal/discovery/discovery.go
// Package discovery 實作 Home Assistant MQTT Discovery（對應 Python solar/discovery.py）。
// 每廠 3 個 summary sensor、每區 6 個 sensor；發佈 retained JSON config，
// HA 自動建立/移除 entity。
package discovery

import (
	"encoding/json"
	"fmt"
	"slices"
	"strings"

	"solar_mqtt_go/internal/scraper"
)

// Publisher 發佈 callback（topic, payload, retain）。
type Publisher func(topic, payload string, retain bool) bool

// devicePayload 對應 Python _device()。
func devicePayload(factoryID string) map[string]any {
	return map[string]any{
		"identifiers":  []string{fmt.Sprintf("ez_solar_%s", factoryID)},
		"name":         fmt.Sprintf("EZ-Solar %s", factoryID),
		"manufacturer": "Joseph-Tech / 國瑞汽車",
		"model":        "EZ-Solar Dashboard",
	}
}

// sensorConfig 對應 Python _sensor_config()；key 順序與 Python dict 一致。
type sensorConfig struct {
	Name          string         `json:"name"`
	UniqueID      string         `json:"unique_id"`
	ObjectID      string         `json:"object_id"`
	StateTopic    string         `json:"state_topic"`
	ValueTemplate string         `json:"value_template"`
	Device        map[string]any `json:"device"`

	Unit      string `json:"unit_of_measurement,omitempty"`
	DeviceCls string `json:"device_class,omitempty"`
	StateCls  string `json:"state_class,omitempty"`
	Icon      string `json:"icon,omitempty"`
}

func sensorConfigNew(factoryID, objectID, name, stateTopic, unit, deviceClass, stateClass string) sensorConfig {
	return sensorConfig{
		Name:          name,
		UniqueID:      objectID,
		ObjectID:      objectID,
		StateTopic:    stateTopic,
		ValueTemplate: "{{ value_json.value }}",
		Device:        devicePayload(factoryID),
		Unit:          unit,
		DeviceCls:     deviceClass,
		StateCls:      stateClass,
	}
}

// zoneMetric 對應 Python metrics 表。
type zoneMetric struct {
	suffix    string
	zh        string
	topicKey  string
	unit      string
	deviceCls string
	stateCls  string
}

var zoneMetrics = []zoneMetric{
	{"power", "功率", "power_kw", "kW", "power", "measurement"},
	{"today", "今日發電", "today_kwh", "kWh", "energy", "total_increasing"},
	{"month", "本月發電", "month_mwh", "MWh", "energy", "total_increasing"},
	{"total", "累計發電", "total_mwh", "MWh", "energy", "total_increasing"},
	{"capacity", "裝置容量", "capacity_kwp", "kWp", "", "measurement"},
	{"hours", "今日有效時數", "today_hours", "h", "", "measurement"},
}

// marshalConfig 序列化（ensure_ascii=False 對齊：不跳脫中文與 HTML 字元）。
func marshalConfig(c sensorConfig) string {
	var sb strings.Builder
	enc := json.NewEncoder(&sb)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(c); err != nil {
		return "{}"
	}
	return strings.TrimSuffix(sb.String(), "\n")
}

// PublishFactoryDiscovery 發佈該廠所有 discovery config，回傳發佈數量。
func PublishFactoryDiscovery(publish Publisher, haPrefix, prefix, factoryID string, zones []scraper.Zone) int {
	base := prefix + "/" + factoryID

	type entry struct {
		objectID string
		cfg      sensorConfig
	}
	var entries []entry

	// Summary
	entries = append(entries,
		entry{
			fmt.Sprintf("ez_solar_%s_total_power", factoryID),
			sensorConfigNew(factoryID,
				fmt.Sprintf("ez_solar_%s_total_power", factoryID),
				fmt.Sprintf("%s 總功率", factoryID),
				base+"/total_power_kw", "kW", "power", "measurement"),
		},
		entry{
			fmt.Sprintf("ez_solar_%s_today_energy", factoryID),
			sensorConfigNew(factoryID,
				fmt.Sprintf("ez_solar_%s_today_energy", factoryID),
				fmt.Sprintf("%s 今日發電", factoryID),
				base+"/today_mwh", "MWh", "energy", "total_increasing"),
		},
		entry{
			fmt.Sprintf("ez_solar_%s_month_energy", factoryID),
			sensorConfigNew(factoryID,
				fmt.Sprintf("ez_solar_%s_month_energy", factoryID),
				fmt.Sprintf("%s 本月發電", factoryID),
				base+"/month_mwh", "MWh", "energy", "total_increasing"),
		},
	)

	// Per-zone
	for _, z := range zones {
		zid := z.ZoneID
		zname := z.Name
		if zname == "" {
			zname = fmt.Sprintf("Zone%d", zid)
		}
		ztopic := fmt.Sprintf("%s/zone/%d", base, zid)
		objPrefix := fmt.Sprintf("ez_solar_%s_zone%d", factoryID, zid)

		for _, m := range zoneMetrics {
			entries = append(entries, entry{
				fmt.Sprintf("%s_%s", objPrefix, m.suffix),
				sensorConfigNew(factoryID,
					fmt.Sprintf("%s_%s", objPrefix, m.suffix),
					fmt.Sprintf("%s %s %s", factoryID, zname, m.zh),
					ztopic+"/"+m.topicKey, m.unit, m.deviceCls, m.stateCls),
			})
		}
	}

	for _, e := range entries {
		topic := fmt.Sprintf("%s/sensor/%s/config", haPrefix, e.objectID)
		publish(topic, marshalConfig(e.cfg), true)
	}

	fmt.Printf("[%s] HA Discovery 已發佈 %d 個 entity\n", factoryID, len(entries))
	return len(entries)
}

// RemoveFactoryDiscovery 清除該廠所有 discovery（空 retained payload → HA 移除 entity）。
func RemoveFactoryDiscovery(publish Publisher, haPrefix, factoryID string, zoneIDs []int) {
	ids := []string{
		fmt.Sprintf("ez_solar_%s_total_power", factoryID),
		fmt.Sprintf("ez_solar_%s_today_energy", factoryID),
		fmt.Sprintf("ez_solar_%s_month_energy", factoryID),
	}
	for _, zid := range zoneIDs {
		for _, suffix := range []string{"power", "today", "month", "total", "capacity", "hours"} {
			ids = append(ids, fmt.Sprintf("ez_solar_%s_zone%d_%s", factoryID, zid, suffix))
		}
	}
	for _, id := range ids {
		publish(fmt.Sprintf("%s/sensor/%s/config", haPrefix, id), "", true)
	}
}

// ── Tracker：service 層使用的重發邏輯（對應 Python _ha_discovery_done / _last_zone_ids） ──

// Tracker 追蹤 discovery 已發佈狀態。
type Tracker struct {
	haPrefix  string
	prefix    string
	factoryID string

	lastZoneIDs []int
	done        bool
	pending     bool // prefix 變更後待重發
}

// NewTracker 建立 tracker。
func NewTracker(haPrefix, prefix, factoryID string) *Tracker {
	return &Tracker{haPrefix: haPrefix, prefix: prefix, factoryID: factoryID}
}

// MarkPrefixChanged 標記 prefix 已變更：不立即發佈，待下一個成功 round 重發。
func (t *Tracker) MarkPrefixChanged() {
	t.pending = true
}

// Pending 回傳是否有待重發的 discovery。
func (t *Tracker) Pending() bool {
	return t.pending
}

// MaybePublish 於每個成功 round 呼叫。首次、zone 清單變動（新增/刪除/reorder）、
// 或 prefix 變更後的下一輪才會發佈。
func (t *Tracker) MaybePublish(publish Publisher, prefix string, zones []scraper.Zone) {
	if prefix != t.prefix {
		t.prefix = prefix
		t.pending = true
	}
	ids := make([]int, 0, len(zones))
	for _, z := range zones {
		ids = append(ids, z.ZoneID)
	}
	if t.done && !t.pending && slices.Equal(ids, t.lastZoneIDs) {
		return
	}
	PublishFactoryDiscovery(publish, t.haPrefix, t.prefix, t.factoryID, zones)
	t.done = true
	t.pending = false
	t.lastZoneIDs = ids
}
